keps/sig-network/5343-kube-proxy-backends/README.md

```diff
@@ -0,0 +1,824 @@
+
+<!--
+**Note:** This checklist is iterative and should be reviewed and updated every time this enhancement is being considered for a milestone.
+-->
+
+[kubernetes.io]: https://kubernetes.io/
+[kubernetes/enhancements]: https://git.k8s.io/enhancements
+[kubernetes/kubernetes]: https://git.k8s.io/kubernetes
+[kubernetes/website]: https://git.k8s.io/website
+
+## Summary
+
+With [KEP-3866] going GA, there are now 4 supported kube-proxy
+backends in-tree (`iptables`, `ipvs`, `nftables`, and `winkernel`).
+The nftables KEP noted that, as future work, we should figure if/how
+we can deprecate either or both of the older Linux backends. There has
+also been sporadic discussion about deprecating the `winkernel`
+backend or moving it out of tree, since it is not clear that it
+benefits from being in-tree.
+
+Meanwhile, an earlier KEP, [KEP-3786], had proposed moving most of the
+shared kube-proxy code into the kube-proxy staging repo
+(`k8s.io/kube-proxy`) so that it could be more easily shared by
+out-of-tree service proxy implementations. That KEP was never merged,
+though there was general agreement with the high-level idea.
+
+The exact goals of this KEP are not finalized, but in general, the
+plan is to figure out what to do with kube-proxy backends.
+
+[KEP-3866]: https://github.com/kubernetes/enhancements/blob/master/keps/sig-network/3866-nftables-proxy/README.md
+[KEP-3786]: https://github.com/kubernetes/enhancements/pull/3788
+
+## Motivation
+
+### Goals
+
+(These are kind of "Meta-Goals", since we still need to decide exactly
+what the Goals of the KEP are.)
+
+- Figure out the situation around "the default kube-proxy backend".
+
+- Figure out a plan for deprecating the `ipvs` backend.
+
+- Figure out (in conjunction with SIG Windows) what we want to do with
+  the `winkernel` backend.
+
+Assuming we end up deciding we want to move any proxy implementation
+out of tree, we will likely want to:
+
+- Refactor and redesign the kube-proxy "library" code
+  (`k8s.io/kubernetes/pkg/proxy`) to make it more usable and more
+  maintainable, and to fix existing bugs.
+
+- Move most of `k8s.io/kubernetes/pkg/proxy/` (except for the backend
+  implementations) into the kube-proxy staging repo,
+  `k8s.io/kube-proxy`, so it can be shared with out-of-tree backends.
+
+### Non-Goals
+
+- We do not want to do anything that sounds like "removing kube-proxy
+  from core kubernetes". In particular:
+
+    - We do not want to remove `cmd/kube-proxy` from
+      `k8s.io/kubernetes`.
+
+    - We do not want to stop building an official Linux kube-proxy
+      container image.
+
+## Proposal
+
+### Determine a timeline for declaring `nftables` to be the "preferred" backend
+
+We expect that *eventually* the `nftables` backend will be a better
+choice for everyone than the `iptables` backend. However, at the
+moment, it requires a new-ish kernel, which may possibly be too new
+for some users.
```

Comment:
How do we determine when the new-ish kerrnel is old enough to be generally available?

Reply:
Oh, the next paragraph answers this 🤦